### PR TITLE
Merge Feature branch with develop branch

### PR DESCRIPTION
**Pull request by ershadul1**
### Project Specification
Created methods which are similar to Ruby built-in Enumerable methods
- my_each
- my_each_with_index
- my_select
- my_all?
- my_any?
- my_none?
- my_count
- my_map
- my_inject

**Checked these methods for different cases**
- Used rubocop to check for linter errors
- We made sure that the methods return the same result as built-in enumerable methods
- Our methods supports Regexp, class, Range.

### CODE REVIEW

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+      for i in 0..length - 1 do
+        yield keys[i], values[i]
+      end
+    else
+      for i in 0..to_a.length - 1 do
+        yield to_a[i]
+      end
+    end
+    self
+  end
```
Comment (christianotieno) [inline main.rb]: # my_each
- [x] `my_each` when a block is given when self is a hash behaves like #each calls the given block once for each element in self

     Failure/Error: expect(my_each_output).to eq(each_output)

       expected: "[:a, 1][:b, 2][:c, 3][:d, 4][:e, 5]"
            got: "abcde"

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+      for i in 0..length - 1 do
+        yield keys[i], values[i], i
+      end
+    else
+      for i in 0..to_a.length - 1 do
+        yield to_a[i], i
+      end
+    end
+    self
+  end
```
Comment (christianotieno) [inline main.rb]: # my_each_with_index
- [x] `my_each_with_index` when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self

     Failure/Error: expect(my_each_with_index_output).to eq(each_with_index_output)

       expected: "Num: [:a, 1], idx: 0\nNum: [:b, 2], idx: 1\nNum: [:c, 3], idx: 2\nNum: [:d, 4], idx: 3\nNum: [:e, 5], idx: 4\n"
            got: "Num: a, idx: 1\nNum: b, idx: 2\nNum: c, idx: 3\nNum: d, idx: 4\nNum: e, idx: 5\n"

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+          counter = false unless i == arg[0]
+        end
+      end
+    else
+      my_each do |i|
+        counter = false unless i
+      end
+    end
+    counter
+  end
```
Comment (christianotieno) [inline main.rb]: # my_all?
- [x] `my_all?` when a Regex is passed as an argument returns true if all of the collection matches the Regex

     Failure/Error: expect(words.my_all?(/d/)).to be words.all?(/d/)

       expected true
            got false

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+          return true if i == arg[0]
+        end
+      end
+    else
+      my_each do |i|
+        return true if i
+      end
+    end
+    counter
+  end
```
Comment (christianotieno) [inline main.rb]: # my_any?
- [x] `my_any?` when a Regex is passed as an argument returns true if any of the collection matches the Regex

     Failure/Error: expect(words.my_any?(/d/)).to be words.any?(/d/)

       expected true
            got false

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+          counter = false if i == arg[0]
+        end
+      end
+    else
+      my_each do |i|
+        counter = false if i
+      end
+    end
+    counter
+  end
```
Comment (christianotieno) [inline main.rb]: # my_none?
   
- [x] `my_none?` when a Regex is passed as an argument returns false if any of the collection matches the Regex

     Failure/Error: expect(words.my_none?(/d/)).to be words.none?(/d/)

       expected false
            got true

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+          map_val << i if i == arg[0]
+        end
+      end
+    else
+      my_each do |i|
+        map_val << i if i
+      end
+    end
+    map_val
+  end
```
Comment (christianotieno) [inline main.rb]: # my_map
- [x] `my_map` returns a new array with the results of running a given block once for every element in enum.
     Failure/Error: expect(array.my_map { |num| num < 10 }).to eq(array.map { |num| num < 10 })

       expected: [true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true...rue, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true]
            got: [1, 0, 3, 6, 8, 1, 8, 6, 3, 7, 5, 6, 2, 1, 0, 0, 4, 4, 3, 4, 4, 2, 3, 1, 6, 5, 8, 6, 3, 8, 1, 6, 0, 8..., 4, 6, 8, 5, 8, 8, 2, 8, 6, 5, 2, 8, 0, 5, 3, 3, 2, 8, 0, 6, 2, 8, 1, 8, 8, 0, 2, 8, 7, 4, 7, 0, 1]

##

- [x] `my_map` returns a new array with the results of running a given block once for every element in enum.
     Failure/Error: expect(array.my_map(&block)).to eq array.map(&block)

       expected: [true, false, false, false, true, false, false, false, false, false, false, true, true, false, true, ..., true, true, true, true, false, false, false, false, false, false, false, true, false, true, false]
            got: [2, 0, 1, 3, 1, 3, 1, 2, 1, 2, 1, 1, 3, 2, 3, 3, 0, 1, 0, 1, 3, 0, 3, 1, 1, 2, 1, 3, 3, 2, 2, 2, 2, 1, 2, 3, 1, 1, 1, 2, 1, 1, 0, 2]

##    

- [x] `my_map` returns a new array with the results of running a given block once for every element in enum::range

     Failure/Error: expect(range.my_map(&block)).to eq range.map(&block)

       expected: [false, false, false, false, false, false, false, false, false, false, false, false, false, false, fa...e, false, false, false, false, false, false, false, false, false, false, false, false, false, false]
            got: []

##

- [x] `my_map` returns an Enumerator if no block is given

     Failure/Error: expect(array.my_map).to be_an(Enumerator)

       expected [8, 3, 7, 2, 5, 1, 5, 2, 7, 3, 8, 6, 4, 2, 5, 0, 3, 1, 0, 7, 2, 7, 2, 0, 7, 1, 8, 7, 1, 3, 3, 0, 7, 8..., 2, 8, 6, 1, 8, 6, 7, 1, 3, 3, 7, 2, 5, 5, 0, 0, 3, 4, 1, 1, 7, 6, 1, 4, 2, 8, 7, 4, 7, 4, 4, 4, 8] to be a kind of Enumerator
##

- [x] `my_map` accepts a proc as an argument and returns a new array with the results of running the proc once for every element in enum.

      Failure/Error: expect(array.my_map(&my_proc)).to eq array.map(&my_proc)

        expected: [false, false, false, false, false, false, false, false, false, false, false, false, false, false, fa...e, false, false, false, false, false, false, false, false, false, false, false, false, false, false]
             got: []

##
- [x] `my_map` executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.

      Failure/Error: expect(array.my_map(my_proc) { |num| num < 10 }).to eq array.map(&my_proc)

        expected: [false, false, false, false, false, false, false, false, false, false, false, false, false, false, fa...e, false, false, false, false, false, false, false, false, false, false, false, false, false, false]
             got: []

File: main.rb [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+      end
+    else
+      result = to_a[0]
+      for i in (1..to_a.length - 1) do
+        result = result.send(arg[0], to_a[i])
+      end
+    end
+    result
+  end
+end
```
Comment (christianotieno) [inline main.rb]: # my_inject 
- [x] `my_inject` raises a "LocalJumpError" when no block or argument is given

      Failure/Error: expect { array.my_inject }.to raise_error(LocalJumpError)

        expected LocalJumpError, got #<TypeError: nil is not a symbol nor a string> with backtrace:
##

- [x] `my_inject` when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block

      Failure/Error: expect(array.my_inject(&operation)).to eq array.inject(&operation)

        expected: 345
             got: 351

##

- [x] `my_inject` when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::range

      Failure/Error: actual = range.my_inject { |prod, n| prod * n }

      NoMethodError:
        undefined method `[]' for 5..50:Range

**Review by christianotieno — CHANGES_REQUESTED** [review]
# **CHANGES REQUIRED** :warning:
Hey @ershadul1 & @Maryjanee  :wave:
-
You have done a splendid job with the enumerable project! :tada: Keep it up.
We, however, do have some few changes to carry out before we proceed (see review comment).
##
## NOTE: These are the variables I used in the tests below: 👇 
```
ARRAY_SIZE = 100
LOWEST_VALUE = 0
HIGHEST_VALUE = 9

describe 'enumerables' do
  let(:array) { Array.new(ARRAY_SIZE) { rand(LOWEST_VALUE...HIGHEST_VALUE) } }
  let(:block) { proc { |num| num < (LOWEST_VALUE + HIGHEST_VALUE) / 2 } }
  let(:words) { %w[dog door rod blade] }
  let(:range) { Range.new(5, 50) }
  let(:hash) { { a: 1, b: 2, c: 3, d: 4, e: 5 } }
  let(:numbers) { [1, 2i, 3.14] }
  let!(:array_clone) { array.clone }
```
## Tests Done: 79
## Failures: 14
- If you have any questions, let me know here or on slack and I will be glad to clarify with you. You can also reach me and connect with me on my social media or email. :eyes:
##
- When you have made the specified changes, please submit another code review. Thank you!
Happy coding :smile:
**[Christian Otieno](https://github.com/ChristianOtieno)**
[Twitter](https://twitter.com/iamchrisotieno) [linkedin](https://www.linkedin.com/in/christianotieno/) [GitLab](https://gitlab.com/christianotieno)
---
*P.S. all requested changes not marked by [OPTIONAL] or are not positive comments are **REQUIRED***

**Comment by ershadul1** [comment]
Fixed all the issues the reviewer pointed out to us.

**Review by rOluochKe — APPROVED** [review]
Hello Team
### Well done, you got it right :clap: The project has been approved. Congratulations :boom: :100:
### Merge and submit the project. Good luck with the next one.
- Happy coding!

**Pull request merged**